src/v/storage/segment.cc

```diff
@@ -470,6 +472,31 @@ ss::future<append_result> segment::append(const model::record_batch& b) {
           return ss::make_exception_future<append_result>(index_err);
       });
 }
+
+ss::future<append_result> segment::append(const model::record_batch& b) {
+    if (has_compaction_index() && b.header().attrs.is_transactional()) {
+        // With transactional batches, we do not know ahead of time whether the
+        // batch will be committed or aborted. We may not have this information
+        // during the lifetime of this segment as the batch may be aborted in
+        // the next segment. We mark this index as `incomplete` and rebuild it
+        // later from scratch during compaction.
+        try {
+            co_await ss::do_with(
+              std::exchange(_compaction_index, std::nullopt), [](auto& index) {
```

Comment:
I don't think you need `do_with` here since this method is already a co-routine. You can `std::exchange` the index directly onto a variable on the stack and it will be kept alive.

Reply:
+1